skimage/metrics/_structural_similarity.py

```diff
@@ -194,11 +194,22 @@ def structural_similarity(im1, im2,
         raise ValueError('Window size must be odd.')
 
     if data_range is None:
+        if (np.issubdtype(im1.dtype, np.floating) or
+            np.issubdtype(im2.dtype, np.floating)):
+            raise ValueError(
+                'Since image dtype is floating point, the data_range '
+                'parameter must be specified. Carefully read the '
+                'documentation (including the note). It is recommended '
+                'to always specify the data_range anyway.')
         if im1.dtype != im2.dtype:
-            warn("Inputs have mismatched dtype.  Setting data_range based on "
-                 "im1.dtype.", stacklevel=2)
+            warn("Inputs have mismatched dtype. Setting data_range based on im1.dtype.",
+                    stacklevel=2)
         dmin, dmax = dtype_range[im1.dtype.type]
         data_range = dmax - dmin
+        if np.issubdtype(im1.dtype, np.integer) and (im1.dtype != np.uint8):
+            warn("Setting data_range based on im1.dtype. " +
```

Comment:
Unrelated to this PR, but I'm just frustrated we still have `im1, im2` around... cf. https://github.com/scikit-image/scikit-image/pull/4187/files#r329088276